{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "906abd30-1368-4d2e-b1be-5a842cd81c31",
   "metadata": {},
   "source": [
    "# Looking at location (distance, within TAD) of features (enhancers, chIP peaks) to dysregulated genes\n",
    "\n",
    "## Plots for DE genes:\n",
    "\n",
    "* All applied threshold for WT expression (total counts) \n",
    "    * Additional distributions of up/down/nonsig genes within these boundaries in **20221029_expression_threshold**\n",
    "* Cumulative Distance from TSS of genes to features (enhancers or ChIP peaks)\n",
    "    * Annotated enhancers (from **20221016_dNipBL_proximal_enhancers**)\n",
    "    * ChiP peak data (**20221108_dNipBL_proximal_proteins**)\n",
    "* Distribution (count) of features within a TAD \n",
    "    * up/down/non-reg genes with some number of enhancers within TAD (**20221016_dNipBL_proximal_enhancers**) \n",
    "    * split by basal expression bin (**20221029_expression_threshold**)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "31097cd2-d12c-456b-b938-feaf58dfd9bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ignore FutureWarnings generated by pandas\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import bioframe as bf\n",
    "from gtfparse import read_gtf\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9cf31f07-4d40-4ccb-aa91-8d95f66454a4",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'plotting_histogram_functions'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m/tmp/SLURM_9420886/ipykernel_3251/196345976.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'../transcription_data_analysis'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mplotting_histogram_functions\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mpfn\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mfilter_gene_functions\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mflt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'plotting_histogram_functions'"
     ]
    }
   ],
   "source": [
    "import sys\n",
    " \n",
    "# setting path\n",
    "sys.path.append('../transcription_data_analysis')\n",
    "\n",
    "import plotting_histogram_functions as pfn\n",
    "import filter_gene_functions as flt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da28daa-dbde-46f2-bc51-6d4410970e23",
   "metadata": {},
   "source": [
    "# Loading all data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa3ba52b-85cf-4df8-8aff-16dc8838e989",
   "metadata": {},
   "source": [
    "### Project Files, annotating TSSes, merging average vst-normalized feature counts into results table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb41e3db-78b0-4964-a769-75c81c0bca28",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj = \"/project/fudenber_735/collaborations/karissa_2022/20220812_EA18-1_RNAseq-Analysis_forGeoff/\"\n",
    "# Importing day 1 depletion in ESCs DEGS\n",
    "day1_sigRes = 'EA18.1_ESC_1d-depletion_DESeq2/20220817_EA18-1_resSig_ESC_1d-depletion.csv'\n",
    "vst_normalized_counts = 'EA18.1_ESC_1d-depletion_DESeq2/20220817_EA18-1_ESC-1d_sf-normalized_vst-transformed.csv'\n",
    "WT_samples = ['KHRNA1', 'KHRNA7', 'KHRNA13', 'KHRNA22', 'KHRNA23', 'KHRNA50']\n",
    "\n",
    "# import results table and add normalized feature counts to label non-significant genes\n",
    "day1_res_df = pd.read_csv(proj+day1_sigRes)\n",
    "vst_counts_df = pd.read_csv(proj+vst_normalized_counts).rename(columns={'Unnamed: 0' : 'Geneid'})\n",
    "vst_counts_df['avg_vst_counts'] = vst_counts_df[WT_samples].mean(axis='columns')\n",
    "day1_res_df = day1_res_df.merge(vst_counts_df[['Geneid', 'avg_vst_counts']], how='outer')\n",
    "\n",
    "# Getting TSSEs\n",
    "tss_df = read_gtf(\"/project/fudenber_735/collaborations/karissa_2022/old/RNAseq/STAR_Gencode_alignment/tss_annotions_gencode.vM23.primary_assembly.gtf\")\n",
    "ts_intervals = flt.get_tss_gene_intervals(tss_df)\n",
    "\n",
    "# labelling our dataset\n",
    "day1_res_df = day1_res_df.merge(ts_intervals, left_on='Geneid', right_on='gene_id', how='left')\n",
    "\n",
    "# If could not find a TSS, drop entire row\n",
    "day1_res_df.dropna(subset=['start'], inplace=True)\n",
    "\n",
    "# Remove any genes mapped to chrX before proceeding --> save these for later analysis\n",
    "day1_res_df = day1_res_df[~day1_res_df.chrom.isin(['chrX'])]\n",
    "day1_res_df = bf.sanitize_bedframe(day1_res_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3349870e-1ad2-4b1f-a4fb-cbc81baf93f5",
   "metadata": {},
   "source": [
    "### import feature data: enhancer files, ChIP files, insulation boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd9bbd5-4d70-4617-84a7-338e851f8524",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_folder = '/project/fudenber_735/collaborations/karissa_2022/2022_09_features_for_RNAseq/Published_datasets/'\n",
    "\n",
    "chen_s1 = 'Enhancers_Chen2012_S1_remapped_mm10.bed'\n",
    "whythe_super = 'Super-enhancers_mESCs_(OSN-MED1)_Wythe-Cell-2023_mm10-lifetover.bed'\n",
    "# whythe_reg = 'remapped_enhancers_Whyte_2013_mm9.bed.txt' <-- where is this?\n",
    "\n",
    "enhancers = {'chen_s1' : dataset_folder+chen_s1,\n",
    "             'wythe_super' : dataset_folder+whythe_super}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb9f1278-bbe6-4953-8e14-71655566d8a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Chip-Seq files \n",
    "chip_folder = \"/project/fudenber_735/collaborations/karissa_2022/2022_09_features_for_RNAseq/ChIP-seq_in_WT-parental-E14/\"\n",
    "\n",
    "ctcf = \"CTCF_peaks_called_on_4reps_foundInatLeast2reps_noBlacklist.bed\"\n",
    "rad21 = \"RAD21_peaks_called_on_6reps_foundInatLeast3reps_noBlacklist.bed\"\n",
    "nipbl = \"Nipbl_112.175.197.114.177.196_peaks.xls.bed\"\n",
    "promoter = \"H3K27ac_EA92-97_peaks.xls.bed\"\n",
    "\n",
    "chip_peaks_condition_dict = {'CTCF' : chip_folder+ctcf, 'RAD21' : chip_folder+rad21,\n",
    "                           'NIPBL' : chip_folder+nipbl, 'H3K27ac' : chip_folder+promoter}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b55b5f28-7ec2-40a7-a0d7-641364c34d9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load features for boundaries\n",
    "feature_dir = '/project/fudenber_735/GEO/bonev_2017_GSE96107/distiller-0.3.1_mm10/results/coolers/features'\n",
    "insulation_table = pd.read_table(feature_dir+'/bonev2017.HiC_ES.mm10.mapq_30.1000.window_200000.insulation')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e11558ea-34e1-4647-a2dc-c52bab3bbeb2",
   "metadata": {},
   "source": [
    "### Annotate Gene Expression Categories, apply threshold and separate dataset into quantiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85dddbe1-fe86-4a0b-95c2-0e20fa8af1fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the differential expression status for filtering genes\n",
    "day1_res_df = flt.label_DE_status(day1_res_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f04ef68-397e-4be0-8af2-8e0b11fd2356",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set counts threshold to 6.0\n",
    "cutoff = 6.0\n",
    "cut = (day1_res_df['avg_vst_counts'] > cutoff)\n",
    "day1_res_df = day1_res_df[cut]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f9453a-031c-49fa-b0cf-b47f26433479",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Break up the categories into quantiles according ot vst_counts\n",
    "quantiled_df = day1_res_df.copy()\n",
    "\n",
    "quantiled_df = flt.label_quantiles(quantiled_df, quantile_value_col=\"avg_vst_counts\")\n",
    "quantiled_df = flt.label_quantiles(quantiled_df, quantile_value_col=\"avg_vst_counts\", label_subset='up')\n",
    "quantiled_df = flt.label_quantiles(quantiled_df, quantile_value_col=\"avg_vst_counts\", label_subset='down')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2e56eef-f7f0-4ce5-b931-70b56701bd54",
   "metadata": {},
   "source": [
    "# Cumulative Distance Plots\n",
    "\n",
    "* Chen/Whyte Enhancers\n",
    "* All ChIP peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b1c0e1-f5a0-4dda-bff0-f379b41f36d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Label distance to closest enhancer\n",
    "for label, file in enhancers.items():\n",
    "    day1_res_df = flt.label_closest_enhancer(day1_res_df, file, label)\n",
    "\n",
    "# Plot\n",
    "fig, axs = plt.subplots(2,2, figsize=(10, 8))\n",
    "\n",
    "for (enhancer_label, file), axi in zip(enhancers.items(), np.arange(axs.shape[0])):\n",
    "    pfn.plot_distance_histogram(day1_res_df, \n",
    "                                 enhancer_label+'_distance', \n",
    "                                 ax=axs[0, axi],\n",
    "                                 cumulative=True,\n",
    "                                 bins=100,\n",
    "                                 plot_title='log10 distance to closest enhancer in {} \\n set from TSS of differentially genes'.format(enhancer_label))\n",
    "    \n",
    "    pfn.plot_distance_histogram(day1_res_df, \n",
    "                                enhancer_label+'_distance', \n",
    "                                ax=axs[1, axi],\n",
    "                                bins=30)\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34247cb3-f494-48ed-a0df-c279bec0812e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotate df with distance to each peak\n",
    "for (chip, file) in chip_peaks_condition_dict.items():\n",
    "    peaks = flt.get_peak_bioframe(file)\n",
    "    day1_res_df = flt.label_closest_peak(day1_res_df, peaks, chip)\n",
    "    \n",
    "fig, axs = plt.subplots(2, len(chip_peaks_condition_dict), figsize=(20, 8))\n",
    "\n",
    "for (chip_name, file), axi in zip(chip_peaks_condition_dict.items(), np.arange(axs.shape[1])):\n",
    "    pfn.plot_distance_histogram(day1_res_df, \n",
    "                                 chip_name+'_distance', \n",
    "                                 ax=axs[0, axi],\n",
    "                                 cumulative=True,\n",
    "                                 bins=100,\n",
    "                                 plot_title='log10 distance to closest {} peak \\n from TSS of differentially genes'.format(chip_name))\n",
    "    \n",
    "    pfn.plot_distance_histogram(day1_res_df, \n",
    "                                chip_name+'_distance', \n",
    "                                ax=axs[1, axi],\n",
    "                                bins=30)\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74edd5d6-1601-462f-9b19-689769787d3e",
   "metadata": {},
   "source": [
    "# Distribution of features and DEGs within a TAD\n",
    "\n",
    "* Chen/Whyte enhancers: basic DE split\n",
    "* Chen/Whyte enhancers: quantile split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3f4a6c1-7334-418d-a5ee-bde0b18d36ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Defining TADs using midpoints between boundaries\n",
    "tad_mid_df = flt.tad_windows_from_boundaries(insulation_table, take_midpoint=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e9072dc-4a1e-499f-928d-9c95e143ada9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2,2, figsize=(10, 8))\n",
    "fig.suptitle(('Distribution of Dysregulated Genes and '\n",
    "             'enhancers within same TAD in 1day NipBL-depleted mESCs'),\n",
    "             fontsize=16)\n",
    "\n",
    "for (enh_source, file), axi in zip(enhancers.items(), np.arange(axs.shape[0])):\n",
    "\n",
    "    # Annotate number of enhancers within the tad\n",
    "    df = tad_mid_df.copy()\n",
    "    enhancer_df = flt.get_enhancer_bioframe(file)\n",
    "    df['num_enhancers'] = bf.count_overlaps(df, enhancer_df)['count']\n",
    "    \n",
    "    pfn.group_features_by_region(df,\n",
    "                                 day1_res_df,\n",
    "                                 feature_category_colors={\"up\": 'tab:red', \n",
    "                                                      \"down\": 'tab:blue'},\n",
    "                                 plot_title='{} enhancers'.format(enh_source),\n",
    "                                 ax=axs[0, axi]\n",
    "                                )\n",
    "                                  \n",
    "    pfn.distribution_features_by_region(df,\n",
    "                                        day1_res_df,\n",
    "                                        plot_title='',\n",
    "                                        ax=axs[1, axi]\n",
    "                                       )\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "579081e0-5a0f-4c43-aa6d-dd5c56af178c",
   "metadata": {},
   "outputs": [],
   "source": [
    "category_colormap_big = {'up_0.0-0.5' : '#fee5d9',\n",
    "                         'up_0.5-0.75' : '#fcae91',\n",
    "                         'up_0.75-0.95' : '#fb6a4a',\n",
    "                         'up_0.95-1.0' : '#cb181d',\n",
    "\n",
    "                         'down_0.0-0.5' : '#deebf7',\n",
    "                         'down_0.5-0.75' : '#bdd7e7',\n",
    "                         'down_0.75-0.95' : '#6baed6',\n",
    "                         'down_0.95-1.0' : '#2171b5',\n",
    "\n",
    "                         'nonsig_0.0-0.5' : '#d9d9d9',\n",
    "                         'nonsig_0.5-0.75' : '#bdbdbd',\n",
    "                         'nonsig_0.75-0.95' : '#969696',\n",
    "                         'nonsig_0.95-1.0' : '#636363'\n",
    "                        }\n",
    "\n",
    "fig, axs = plt.subplots(2,2, figsize=(10, 8))\n",
    "fig.suptitle(('Distribution of Dysregulated Genes and '\n",
    "             'enhancers within same TAD in 1day NipBL-depleted mESCs'),\n",
    "             fontsize=16)\n",
    "\n",
    "for (enh_source, file), axi in zip(enhancers.items(), np.arange(axs.shape[0])):\n",
    "\n",
    "    # Annotate number of enhancers within the tad\n",
    "    df = tad_mid_df.copy()\n",
    "    \n",
    "    enhancer_df = flt.get_enhancer_bioframe(file)\n",
    "    df['num_enhancers'] = bf.count_overlaps(df, enhancer_df)['count']\n",
    "    \n",
    "    pfn.group_features_by_region(df,\n",
    "                                 quantiled_df,\n",
    "                                 feature_category_colors=category_colormap_big,\n",
    "                                 plot_title='{} enhancers'.format(enh_source),\n",
    "                                 ax=axs[0, axi]\n",
    "                                )\n",
    "                                  \n",
    "    pfn.distribution_features_by_region(df,\n",
    "                                        quantiled_df,\n",
    "                                        plot_title='',\n",
    "                                        feature_category_colors=category_colormap_big,\n",
    "                                        ax=axs[1, axi]\n",
    "                                       )\n",
    "    \n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bioframe Env",
   "language": "python",
   "name": "bioenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
