{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "99fe7ed2-412d-462b-9bb7-8941333131a0",
   "metadata": {},
   "source": [
    "# Plotting ChIP signal around TSS of DEGs after 24h NipBL while measuring background levels\n",
    "\n",
    "***Purpose:*** To build better understanding of typical background level of ChIP signal in wildType genes, we try splitting the non-significantly regulated genes into different categories based on their basal level of expression (as measured by Counts). This will be used for comparison against up- or down-regulated genes in the 24-hour dNipBL degradation set when averaging over ChIP signal at the TSS. We might expect to find patterns such as genes that are already highly expressed, but are not impacted by degradation of NipBL (i.e. do not appear in up- or down-regulated gene set) correspond to one of the post-degradation average signal curves.\n",
    "\n",
    "***General Process:***\n",
    "- Use vst normalized counts to quantify expression levels for any genes not in the significant results table\n",
    "- Take average of vst_normalized counts across all wild-type samples in this set\n",
    "- Split the counts into quartiles: bottom 25% counts, 25-50%, 50-75%, 75-100% for all genes tagged as non-significant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ce26c1b8-1450-44c4-80a7-c40327036bc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotting_heatmap_functions as pfn\n",
    "import filter_gene_functions as flt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7873ab2e-5c2e-40f3-af5f-35911df2b5be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ignore FutureWarnings generated by pandas\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import bbi\n",
    "from gffutils.helpers import asinterval\n",
    "from gtfparse import read_gtf\n",
    "import bioframe as bf\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "from mpl_toolkits.axes_grid1.inset_locator import inset_axes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8c0c34b-0074-4aca-ba33-0acafc632083",
   "metadata": {},
   "source": [
    "# Prepare Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "37fbbb27-050f-4eaf-b63c-e020ff1eccc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj = \"/project/fudenber_735/collaborations/karissa_2022/20220812_EA18-1_RNAseq-Analysis_forGeoff/\"\n",
    "\n",
    "# Importing day 1 depletion in ESCs DEGS\n",
    "day1_sigRes = 'EA18.1_ESC_1d-depletion_DESeq2/20220817_EA18-1_resSig_ESC_1d-depletion.csv'\n",
    "\n",
    "# Sample count data for the non-significant results\n",
    "normalized_counts = 'EA18.1_ESC_1d-depletion_DESeq2/20220817_EA18-1_ESC-1d_sf-normalized.csv'\n",
    "vst_normalized_counts = 'EA18.1_ESC_1d-depletion_DESeq2/20220817_EA18-1_ESC-1d_sf-normalized_vst-transformed.csv'\n",
    "feature_counts = '20220816_featureCounts.csv'\n",
    "\n",
    "WT_samples = ['KHRNA1', 'KHRNA7', 'KHRNA13', 'KHRNA22', 'KHRNA23', 'KHRNA50']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "caa76072-21b8-44c9-97ab-695b55b72f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "day1_res_df = pd.read_csv(proj+day1_sigRes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c95ed6fa-8f48-441f-aec4-cc7b0c984805",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1619, 5)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "day1_res_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edc7cced-231e-46cc-aed5-87321e1d824c",
   "metadata": {},
   "source": [
    "### Merging significant results and counts to get a complete dataset\n",
    "\n",
    "* load the raw feature counts to collect the full set of non-significantly DE genes\n",
    "* average over all WT features, for comparison\n",
    "* get averaged counts using the vst_normalized counts --> use these for analyzing relative abundance\n",
    "* Create a full df for the set of genes with:\n",
    "   - results values from `day1_sigRes`, \n",
    "   the full gene set from `normalized_counts`, values of normalized counts from `vst_normalized_counts`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "985ff5a5-44e0-4ed1-87ed-7bd57d5a1451",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raw feature counts shape:  (55421, 110)\n",
      "vst normalized feature counts shape:  (36446, 14)\n"
     ]
    }
   ],
   "source": [
    "# import table of raw feature counts and calculate average\n",
    "feat_counts_df = pd.read_csv(proj+feature_counts).rename(columns={'Unnamed: 0' : 'Geneid'})\n",
    "feat_counts_df['avg'] = feat_counts_df[WT_samples].mean(axis='columns')\n",
    "print('raw feature counts shape: ', str(feat_counts_df.shape))\n",
    "\n",
    "# import table of normalized feature counts and calculate average\n",
    "vst_counts_df = pd.read_csv(proj+vst_normalized_counts).rename(columns={'Unnamed: 0' : 'Geneid'})\n",
    "vst_counts_df['avg'] = vst_counts_df[WT_samples].mean(axis='columns')\n",
    "print('vst normalized feature counts shape: ', str(vst_counts_df.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4f461330-34a3-4808-990b-22bb06c67e56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(55421, 123)\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "feat_counts_df = feat_counts_df.merge(vst_counts_df, on='Geneid', how='left', suffixes=('_counts', '_vst_counts'))\n",
    "feat_counts_df['avg_vst_counts'].fillna(feat_counts_df['avg_counts'], inplace=True)\n",
    "\n",
    "print(feat_counts_df.shape)\n",
    "print(feat_counts_df['avg_vst_counts'].isna().sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f182d95b-1daf-497c-9e39-5a9c3075e953",
   "metadata": {},
   "source": [
    "Note: by inspection, can see that avg_vst_counts == NaN wherever avg_counts == 0. Filling missing avg_vst_counts with avg_counts is effectively replacing all NaNs with zeros"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a5f22f40-5250-4ce1-a324-c8984770314b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add average normalized counts value to results df\n",
    "day1_res_df = day1_res_df.merge(feat_counts_df[['Geneid', 'avg_vst_counts', 'avg_counts']], on='Geneid', how='outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1fad5c0-d170-480a-8e99-7c1a21b8828b",
   "metadata": {},
   "source": [
    "### Adding TSS coordinate locations from genomic annotation file\n",
    "\n",
    "Getting the TSS for the genes in this set\n",
    "\n",
    "Remove any genes mapped to chrX before proceeding (save for later analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "61a982db-cba6-4e96-bce3-62bf0f094cdc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'transcript_id', 'gene_type', 'gene_name', 'transcript_type', 'transcript_name', 'level', 'transcript_support_level', 'mgi_id', 'tag', 'havana_gene', 'havana_transcript', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking for 'lost' significant genes: \n",
      "lost:  0\n"
     ]
    }
   ],
   "source": [
    "# Getting TSSEs\n",
    "tss_df = read_gtf(\"/project/fudenber_735/collaborations/karissa_2022/old/RNAseq/STAR_Gencode_alignment/tss_annotions_gencode.vM23.primary_assembly.gtf\")\n",
    "ts_intervals = flt.get_tss_gene_intervals(tss_df)\n",
    "\n",
    "# labelling our dataset\n",
    "day1_res_df = day1_res_df.merge(ts_intervals, left_on='Geneid', right_on='gene_id', how='left')\n",
    "\n",
    "# Checking if any of the significant genes will be lost due to loss of TSS annotation\n",
    "print(\"Checking for 'lost' significant genes: \")\n",
    "ix = np.where(day1_res_df['start'].isna() & (day1_res_df['padj'] < 0.05))\n",
    "print(\"lost: \", str(day1_res_df.iloc[ix].shape[0]))\n",
    "# If could not find a TSS, drop entire row\n",
    "day1_res_df.dropna(subset=['start'], inplace=True)\n",
    "\n",
    "# Remove any genes mapped to chrX before proceeding --> save these for later analysis\n",
    "day1_res_df = day1_res_df[~day1_res_df.chrom.isin(['chrX'])]\n",
    "day1_res_df = bf.sanitize_bedframe(day1_res_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ef530fd-f35b-4853-9daf-d0bd7ba1d35e",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Separating non-significant genes by quartiles of vst counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4351ac6a-c90b-4d1b-b455-68361c4f288a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, label DE status by up, down, or non-significant\n",
    "day1_res_df = flt.label_DE_status(day1_res_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1990b162-316a-491b-967a-5dd5d0e4c04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# label quartiles (with default q_array = [0, .5, .75, .95, 1.])\n",
    "day1_res_df = flt.label_quantiles(day1_res_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddfd57ef-44ef-4a0b-9b59-10d6d2e26390",
   "metadata": {},
   "source": [
    "### Import ChIP-seq files and measure over TSS intervals "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "94483dbf-32bd-4eba-8f69-83db2597d85e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Chip-Seq files \n",
    "chip_folder = \"/project/fudenber_735/collaborations/karissa_2022/2022_09_features_for_RNAseq/ChIP-seq_in_WT-parental-E14/\"\n",
    "\n",
    "ctcf_new = \"CTCF_E14_RSC13new-22-37-60_average.bw\"\n",
    "rad21 = \"RAD21_E14_RSC12new-21-36-59-74_average.bw\"\n",
    "nipbl = \"NIPBL_E14_EA112-EA175_average.bw\"\n",
    "ring1b = \"RING1B_E14_RSC24-39-62_average.bw\"\n",
    "promoter = \"H3K27Ac_mESCs_EA92-EA94_average.bw\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dbeadccc-1a83-4b54-afdb-39db84d0e566",
   "metadata": {},
   "outputs": [],
   "source": [
    "chip_seq_condition_dict = {'CTCF' : chip_folder+ctcf_new, 'RAD21' : chip_folder+rad21,\n",
    "                           'NIPBL' : chip_folder+nipbl, 'RING1B' : chip_folder+ring1b,\n",
    "                           'H3K27ac' : chip_folder+promoter}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "58a0fa9d-31e8-43a8-b117-17bbff0e41d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate ChIP stack-up matrices for all proteins of interest, for plotting later:\n",
    "chip_seq_condition_matrix_dict = pfn.generate_signal_matrix_parallel(day1_res_df, chip_seq_condition_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "37b1ab0e-7807-4bab-bb4c-44f810283273",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['down' 'nonsig_0.0-0.5' 'nonsig_0.5-0.75' 'nonsig_0.75-0.95'\n",
      " 'nonsig_0.95-1.0' 'up']\n"
     ]
    }
   ],
   "source": [
    "categories = np.unique(day1_res_df['DE_status'])\n",
    "print(categories)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e4e44d5b-0728-4e51-a660-0ed48e4fd187",
   "metadata": {},
   "outputs": [],
   "source": [
    "category_colormap = {'up' : 'r',\n",
    "                     'down' : 'b',\n",
    "                     'nonsig_0.0-0.5' : '#d9d9d9',\n",
    "                     'nonsig_0.5-0.75' : '#bdbdbd',\n",
    "                     'nonsig_0.75-0.95' : '#969696',\n",
    "                     'nonsig_0.95-1.0' : '#636363'\n",
    "                    }"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b77ea91-045d-4db2-bd0e-8b29cddfe41a",
   "metadata": {},
   "source": [
    "# Generate average signal plots for all proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2466511c-d6e9-4f64-aaec-07dddcfad7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#last_ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d1ab98ad-bbb1-4cbb-8026-3aea7193a109",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gridspec' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [18], line 11\u001b[0m\n\u001b[1;32m      7\u001b[0m fig \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39mfigure(figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m20\u001b[39m, \u001b[38;5;241m14\u001b[39m))\n\u001b[1;32m      8\u001b[0m fig\u001b[38;5;241m.\u001b[39msuptitle((\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChIP signal around TSS of DEGs \u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      9\u001b[0m              \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdetected after 1 day NipBL depletion in ESCs\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[1;32m     10\u001b[0m              fontsize\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m16\u001b[39m)\n\u001b[0;32m---> 11\u001b[0m gs0 \u001b[38;5;241m=\u001b[39m \u001b[43mgridspec\u001b[49m\u001b[38;5;241m.\u001b[39mGridSpec(\u001b[38;5;241m2\u001b[39m, \u001b[38;5;241m1\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m2\u001b[39m\u001b[38;5;241m*\u001b[39m(num_conditions\u001b[38;5;241m+\u001b[39mnum_deg_sets), \n\u001b[1;32m     12\u001b[0m                               figure\u001b[38;5;241m=\u001b[39mfig, \n\u001b[1;32m     13\u001b[0m                               width_ratios\u001b[38;5;241m=\u001b[39mwidths,\n\u001b[1;32m     14\u001b[0m                        height_ratios\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m.5\u001b[39m])\n\u001b[1;32m     16\u001b[0m gs0\u001b[38;5;241m.\u001b[39mupdate(wspace\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.45\u001b[39m, hspace\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.2\u001b[39m)\n\u001b[1;32m     17\u001b[0m cax1 \u001b[38;5;241m=\u001b[39m fig\u001b[38;5;241m.\u001b[39madd_subplot(gs0[\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m0\u001b[39m])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gridspec' is not defined"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 2000x1400 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "num_conditions = len(chip_seq_condition_dict)\n",
    "num_deg_sets = 1\n",
    "widths = [0.1, .05, 0.5]*num_deg_sets+[1.25, 0.1]*num_conditions\n",
    "\n",
    "\n",
    "plt.rcParams['font.size'] =12\n",
    "fig = plt.figure(figsize=(20, 14))\n",
    "fig.suptitle(('ChIP signal around TSS of DEGs '\n",
    "             'detected after 1 day NipBL depletion in ESCs'),\n",
    "             fontsize=16)\n",
    "gs0 = gridspec.GridSpec(2, 1 + 2*(num_conditions+num_deg_sets), \n",
    "                              figure=fig, \n",
    "                              width_ratios=widths,\n",
    "                       height_ratios=[1, .5])\n",
    "\n",
    "gs0.update(wspace=0.45, hspace=0.2)\n",
    "cax1 = fig.add_subplot(gs0[0, 0])\n",
    "ax1 = fig.add_subplot(gs0[0, 2])\n",
    "\n",
    "pfn.plot_category_heatmap(day1_res_df, \n",
    "                          plot_title=\"\"\"DEG expr\"\"\",\n",
    "                          ax=ax1,\n",
    "                          cax=cax1,\n",
    "                          color_tick_percentiles=[5,25,50,75,95],\n",
    "                          )\n",
    "last_ax = 2\n",
    "for cond, chip_file in chip_seq_condition_dict.items():\n",
    "    \n",
    "    ax2 = fig.add_subplot(gs0[0, last_ax+1], sharey=ax1)\n",
    "    cax2 = ax2.inset_axes((1.05,0.0,.1,1.0))\n",
    "    \n",
    "    ax3 = fig.add_subplot(gs0[1, last_ax+1], sharex=ax2)\n",
    "    \n",
    "    last_ax=last_ax+2\n",
    "\n",
    "\n",
    "    matrix = chip_seq_condition_matrix_dict[cond]\n",
    "\n",
    "\n",
    "    if (last_ax==gs0.ncols-1):\n",
    "        add_legend=True\n",
    "        colorbar_label ='ChIP signal'\n",
    "    else:\n",
    "        add_legend=False\n",
    "        colorbar_label =None\n",
    "\n",
    "    pfn.plot_binned_signal_heatmap(day1_res_df, \n",
    "                                   matrix, \n",
    "                                   plot_title=(\"{} ChIP signal\".format(cond)),\n",
    "                                   ax=ax2,\n",
    "                                   cax=cax2,\n",
    "                                   color_tick_percentiles=[50,95,99],\n",
    "                                   colorbar_label = colorbar_label\n",
    "                                  )\n",
    "    plt.setp(ax2.get_yticklabels(), visible=False)\n",
    "\n",
    "    pfn.plot_avg_signal(day1_res_df, \n",
    "                        matrix, \n",
    "                        'Avg {} ChIP'.format(cond), \n",
    "                        ax=ax3, \n",
    "                        category_colors=category_colormap,\n",
    "                        DE_value_col='log2Fold_Change',\n",
    "                        agg_key='DE_status', \n",
    "                        window_size=1000, \n",
    "                        nbins=40,\n",
    "                        add_legend=add_legend,\n",
    "                       )\n",
    "    ylim = plt.gca().get_ylim()\n",
    "    plt.ylim([-.1, ylim[1]*1.05])\n",
    "    \n",
    "plt.savefig('20221012_1day_dNipBL_vs_ChIP.png')\n",
    "plt.savefig('20221012_1day_dNipBL_vs_ChIP.pdf')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "open2c",
   "language": "python",
   "name": "open2c"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
